pinot-core/src/main/java/org/apache/pinot/core/io/util/PinotDataBitSetV2.java

```diff
@@ -24,109 +24,58 @@
-   * reads and also takes care of the small byte-sized window of unaligned read.
-   * @param startIndex start docId
-   * @param length length
-   * @param out out array to store the unpacked integers
-   */
-  public abstract void readInt(long startIndex, int length, int[] out);
-
-  /**
-   * Decode integers for an array of indexes which is not necessarily
-   * contiguous. So there could be gaps in the array:
-   * e.g: [1, 3, 7, 9, 11, 12]
-   * The actual read is done by the previous API since that is efficient
-   * as it exploits contiguity and uses vectorization. However, since
-   * the out[] array has to be correctly populated with the unpacked integer
-   * for each index, a post-processing step is needed after the bulk contiguous
-   * read to correctly set the unpacked integer into the out array throwing away
-   * the unnecessary values decoded as part of contiguous read
-   * @param docIds index array
-   * @param docIdsStartIndex starting index in the docIds array
-   * @param length length to read (number of docIds to read in the array)
-   * @param out out array to store the decoded integers
-   * @param outpos starting index in the out array
-   */
-  public void readInt(int[] docIds, int docIdsStartIndex, int length, int[] out, int outpos) {
-    int startDocId = docIds[docIdsStartIndex];
-    int endDocId = docIds[docIdsStartIndex + length - 1];
-    int[] dictIds = THREAD_LOCAL_DICT_IDS.get();
-    // do a contiguous bulk read
-    readInt(startDocId, endDocId - startDocId + 1, dictIds);
-    out[outpos] = dictIds[0];
-    // set the unpacked integer correctly. this is needed since there could
-    // be gaps and some decoded values may have to be thrown/ignored.
-    for (int i = 1; i < length; i++) {
-      out[outpos + i] = dictIds[docIds[docIdsStartIndex + i] - startDocId];
-    }
-  }
+public abstract class PinotDataBitSetV2 extends BasePinotBitSet implements PinotBitSet, Closeable {
+  static final int MAX_VALUES_UNPACKED_SINGLE_ALIGNED_READ = 32;
 
-  public static PinotDataBitSetV2 createBitSet(PinotDataBuffer pinotDataBuffer, int numBitsPerValue) {
-    switch (numBitsPerValue) {
-      case 1:
-        return new Bit1Encoded(pinotDataBuffer, numBitsPerValue);
-      case 2:
-        return new Bit2Encoded(pinotDataBuffer, numBitsPerValue);
-      case 4:
-        return new Bit4Encoded(pinotDataBuffer, numBitsPerValue);
-      case 8:
-        return new Bit8Encoded(pinotDataBuffer, numBitsPerValue);
-      case 16:
-        return new Bit16Encoded(pinotDataBuffer, numBitsPerValue);
-      case 32:
-        return new RawInt(pinotDataBuffer, numBitsPerValue);
-      default:
-        throw new UnsupportedOperationException(numBitsPerValue + "not supported by PinotDataBitSetV2");
-    }
+  public PinotDataBitSetV2(PinotDataBuffer dataBuffer, int numBits) {
+    super(dataBuffer, numBits);
   }
 
   public static class Bit1Encoded extends PinotDataBitSetV2 {
+    // grab a final local reference to avoid
```

Comment:
Do you observe performance degradation on this? I think hotspot can definitely handle this. Keeping an extra reference seems like over-optimization to me.